internal/bootstrap/deployer.go

```diff
@@ -0,0 +1,498 @@
+		objectStore:        config.ObjectStore,
+		constraints:        config.Constraints,
+		controllerConfig:   config.ControllerConfig,
+		newCharmRepo:       config.NewCharmRepo,
+		newCharmDownloader: config.NewCharmDownloader,
+		charmhubHTTPClient: config.CharmhubHTTPClient,
+		charmhubURL:        config.CharmhubURL,
+		channel:            config.Channel,
+		loggerFactory:      config.LoggerFactory,
+		logger:             config.LoggerFactory.Child("deployer"),
+	}
+}
+
+// ControllerCharmArch returns the architecture used for deploying the
+// controller charm.
+func (b *baseDeployer) ControllerCharmArch() string {
+	arch := corearch.DefaultArchitecture
+	if b.constraints.HasArch() {
+		arch = *b.constraints.Arch
+	}
+	return arch
+}
+
+// DeployLocalCharm deploys the controller charm from the local charm
+// store.
+func (b *baseDeployer) DeployLocalCharm(ctx context.Context, arch string, base corebase.Base) (*charm.URL, *corecharm.Origin, error) {
+	controllerCharmPath := filepath.Join(b.dataDir, "charms", bootstrap.ControllerCharmArchive)
+	_, err := os.Stat(controllerCharmPath)
+	if os.IsNotExist(err) {
+		return nil, nil, errors.NotFoundf(controllerCharmPath)
+	}
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	curl, err := addLocalControllerCharm(ctx, b.objectStore, b.charmUploader, base, controllerCharmPath)
+	if err != nil {
+		return nil, nil, errors.Annotatef(err, "cannot store controller charm at %q", controllerCharmPath)
+	}
+	b.logger.Debugf("Successfully deployed local Juju controller charm")
+	origin := corecharm.Origin{
+		Source: corecharm.Local,
+		Type:   "charm",
+		Platform: corecharm.Platform{
+			Architecture: arch,
+			OS:           base.OS,
+			Channel:      base.Channel.String(),
+		},
+	}
+	return curl, &origin, nil
+}
+
+// DeployCharmhubCharm deploys the controller charm from charm hub.
+func (b *baseDeployer) DeployCharmhubCharm(ctx context.Context, arch string, base corebase.Base) (string, *corecharm.Origin, error) {
+	model, err := b.stateBackend.Model()
+	if err != nil {
+		return "", nil, errors.Trace(err)
+	}
+
+	charmRepo, err := b.newCharmRepo(services.CharmRepoFactoryConfig{
+		LoggerFactory:      b.loggerFactory,
+		CharmhubHTTPClient: b.charmhubHTTPClient,
+		StateBackend:       b.charmUploader,
+		ModelBackend:       model,
+	})
+	if err != nil {
+		return "", nil, errors.Trace(err)
+	}
+
+	var curl *charm.URL
+	if b.charmhubURL == "" {
+		curl = charm.MustParseURL(controllerCharmURL)
+	} else {
+		curl = charm.MustParseURL(b.charmhubURL)
+	}
+	if err != nil {
+		return "", nil, errors.Trace(err)
+	}
+	origin := corecharm.Origin{
+		Source:  corecharm.CharmHub,
+		Type:    "charm",
+		Channel: &b.channel,
+		Platform: corecharm.Platform{
+			Architecture: arch,
+			OS:           base.OS,
+			Channel:      base.Channel.Track,
+		},
+	}
+
+	// Since we're running on the machine to which the controller charm will be
+	// deployed, we know the exact platform to ask for, not need to review the
```

Comment:
"no need". I'd also refer to _base_ rather than _series_, which we're strangling out.